sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/SymbolUsageCollector.cs

```diff
@@ -0,0 +1,284 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace SonarAnalyzer.Helpers
+{
+    internal class SymbolUsageCollector : CSharpSyntaxWalker
+    {
+        private static readonly ISet<SyntaxKind> IncrementKinds = new HashSet<SyntaxKind>
+        {
+            SyntaxKind.PostIncrementExpression,
+            SyntaxKind.PreIncrementExpression,
+            SyntaxKind.PostDecrementExpression,
+            SyntaxKind.PreDecrementExpression
+        };
+
+        private readonly Func<SyntaxNode, SemanticModel> getSemanticModel;
+        private readonly HashSet<string> removableSymbolNames;
+
+        public HashSet<ISymbol> UsedSymbols { get; } =
+            new HashSet<ISymbol>();
+
+        public HashSet<string> DebuggerDisplayValues { get; } =
+            new HashSet<string>();
+
+        public Dictionary<IPropertySymbol, AccessorAccess> PropertyAccess { get; } =
+            new Dictionary<IPropertySymbol, AccessorAccess>();
+
+        public SymbolUsageCollector(Func<SyntaxTree, bool, SemanticModel> getSemanticModel, HashSet<string> removableSymbolNames)
```

Comment:
Naming: _knownIdentifierNames_ would be better than _removableSymbolNames_ - from the point of view of this class, it doesn't know or care about removability.